ntp-proto/src/packet/mod.rs

```diff
@@ -456,7 +456,87 @@ impl<'a> NtpPacket<'a> {
                     recv_timestamp,
                     clock,
                 )),
-                efdata: Default::default(),
+                efdata: ExtensionFieldData {
+                    authenticated: vec![],
+                    encrypted: vec![],
+                    // Ignore encrypted so as not to accidentaly leak anything
+                    untrusted: input
+                        .efdata
+                        .untrusted
+                        .iter()
+                        .chain(input.efdata.authenticated.iter())
+                        .filter_map(|f| {
+                            if matches!(f, ExtensionField::UniqueIdentifier(_)) {
+                                Some(f.clone())
+                            } else {
+                                None
+                            }
+                        })
+                        .collect(),
+                },
+                mac: None,
+            },
+        }
+    }
+
+    pub fn nts_timestamp_response<C: NtpClock>(
+        system: &SystemSnapshot,
+        input: Self,
+        recv_timestamp: NtpTimestamp,
+        clock: &C,
+        cookie: &DecodedServerCookie,
+        keyset: &KeySet,
+    ) -> Self {
+        match input.header {
+            NtpHeader::V3(_) => unreachable!("NTS shouldn't work with NTPv3"),
+            NtpHeader::V4(header) => NtpPacket {
+                header: NtpHeader::V4(NtpHeaderV3V4::timestamp_response(
+                    system,
+                    header,
+                    recv_timestamp,
+                    clock,
+                )),
+                efdata: ExtensionFieldData {
+                    authenticated: input
+                        .efdata
+                        .authenticated
+                        .iter()
+                        .filter_map(|f| {
+                            if matches!(f, ExtensionField::UniqueIdentifier(_)) {
+                                Some(f.clone())
+                            } else {
+                                None
+                            }
+                        })
+                        .collect(),
+                    encrypted: input
+                        .efdata
+                        .authenticated
+                        .iter()
+                        .chain(input.efdata.encrypted.iter())
+                        .filter_map(|f| {
+                            if let ExtensionField::NtsCookiePlaceholder { cookie_length } = f {
+                                let new_cookie = keyset.encode_cookie(cookie);
+                                if new_cookie.len() > *cookie_length as usize {
+                                    None
+                                } else {
+                                    Some(ExtensionField::NtsCookie(Cow::Owned(new_cookie)))
+                                }
```

Comment:
this logic can be factored out (and hence deduplicated)

Reply:
maybe even move this whole closure into its own function

Reply:
done